### ISSUE TITLE
Shotgun treating Slug as 2-shot magazine

### ISSUE THREAD

**Issue by DeeJayhX** [unitystation/unitystation#3312]
# Issue Template

## Description
Currently, you can only load one slug into the shotgun, fire the shotgun twice (during which, the shotgun ejects regular bullet casings instead of spent slugs), and then eject the slug like an empty magazine.

Each slug should be one shot, the slug should be ejected like a casing when fired (and not eject a bullet casing), the slug should look blown open at the plastic end when ejected, and you should be able to load at least 4 slugs into the shotgun (this is the BARE minimum, for a super short combat shotgun with just a cap in place of a magazine tube...)

The ammo capacity should probably be as follows for the current shotguns:
Sawn_Shotgun_Ballistic = 4 (1 in the chamber, 3 in the magazine tube)
Shotgun_Ballistic = 4 (1 in the chamber, 3 in the magazine tube)
Cshotgun_Ballistic = 6 (1 in the chamber, 5 in the magazine tube)

This would probably be a good easy/starter issue.

### Steps to Reproduce
Please enter the steps to reproduce the bug or behavior:  

1.  Grab a shotgun
2.  Fire it twice
3.  Notice the bullet casings hitting the floor
4.  Eject the unspent-looking, magical 2 shot slug
5.  Praise the honkmother

**Comment by drneuris** [comment]
just spitballing here, but could we have the slugs be stackable, to maybe like 5-8? I feel it takes TONS of room in inventory to have a reasonable amount of bullets if you decide to go shotty

**Comment by DeeJayhX** [comment]
> just spitballing here, but could we have the slugs be stackable, to maybe like 5-8? I feel it takes TONS of room in inventory to have a reasonable amount of bullets if you decide to go shotty

There's no way for that to be feasible. You couldn't realistically cram more than three 10 or 12 gauge shotgun shells into your pocket and still use them to reload reliably. At that point, what's even the point of making them stackable. Plus, this is exactly what the shotgun bandolier is for.

**Comment by Dropgunner** [comment]
A normal combat shotgun holds 7+1 rounds in his chamber. 2 rounds are definitely not enough

**Comment by DeeJayhX** [comment]
> A normal combat shotgun holds 7+1 rounds in his chamber. 2 rounds are definitely not enough

@Dropgunner Actually, a normal combat shotgun holds 6+1. Good thing the bandolier only costs 5 pieces of leather, lets you hold almost 3 reloads (since it holds 15 slugs), and divides them by payload type for easy selecting. Shotgun shells do not need to be stack-able. Period.

https://streamable.com/m0dr1

**Comment by Tjarr737** [comment]
I am going to try my hand at this:
I intend to make a subclass InternalMagazineGun of Gun, which uses the regular magazine slot for the currently chambered round and has an additional stack (with possibly a choice of queue) of magazines. The chambered 'magazine' will be despawned on exhaustion with the spawned casing representing the used 'magazine'. The shells as they currently are would then be magazines of capacity one. If we ever wanted a shotgun with a drum afterwards, the drum would have to be a different type of magazine (for example MultiSlug instead of Slug).
The casing is currently generally fixed, with projectile and projectile number being part of the gun definition. This should be changed to allow for multiple types of ammunition for one weapon (for example buckshot and slugs), considering that the damage for regular weapons mainly depends on the ammunition, however the existence of energy weapons (and possibly some other high-tech shenanigans) complicates this, and I might skip this for now.
From what I see, this approach should both require no changes in other parts of the code and still be rather flexible.
Edit: Gun.cs has to be changed at least a bit due to encapsulation (push some over to protected instead of private).
Edit: Due to the workings behind the magazine ejection, it seems more reasonable to have the magazine (in this case the slug) change their own graphics on expenditure, rather than having a separate casing object.

**Comment by drneuris** [comment]
this hurts my brain somehow, might be me, but consider these points
>I intend to make a subclass InternalMagazineGun of Gun, which uses the regular magazine slot for the currently chambered round and has an additional stack (with possibly a choice of queue) of magazines.

why subclass gun when you could have the internal magazine be... a magazine, but non removable
why not just have shells be shells instead and refill magazines instead of being magazines of size1 themselves
with our interaction system it's pretty easy to set up, and it's less pain when you inevitably will have to change it, you say it yourself
> If we ever wanted a shotgun with a drum afterwards, the drum would have to be a different type of magazine (for example MultiSlug instead of Slug).

makes little sense to me.

> There's no way for that to be feasible. You couldn't realistically cram more than three 10 or 12 gauge shotgun shells into your pocket and still use them to reload reliably. At that point, what's even the point of making them stackable. Plus, this is exactly what the shotgun bandolier is for.

what
so you're saying a 100round lmg magazine in a pocket is fine, but a stack of 3-5 shotgun shells is not? 
literally every other inventory system has stackable shotgun shells for this reason, i see nothing wrong with having a few scales available, like 2-3 rounds from pockets or normal inventory slots, three or four times that for a bandolier (just have three slots instead of 12), etc

**Comment by Tjarr737** [comment]
I decided on these choices because the current implementation of guns and magazines does not allow to comfortably follow your propositions:
  1. Magazines are forcefully spawned into guns at first depending on the ammo type. If we wanted to use internal ammo containers we would have define a new ammo type for each.
  2. Guns always allow ejection of the contained magazine. Non-removable is simply not allowed at the moment.
  3. Magazines track a count of ammo instead of contained ammo (different types).

After some consideration I agree, that an internal magazine is probably the better choice for now - I simply chose to avoid doing this at all, because 
 1. I wanted to go for multiple ammo types directly, which is not supported by magazines or guns at all, which actually makes that point kind of silly.
 2. Doing it the previously described way requires very little change to reloading/unloading, as we would still be working on the currently inserted magazine.

I therefore still think making an extra subclass with a forced internal magazine and the ability for pump-action is reasonable. And it is probably a better choice than making some additional reloadable-script which has to be attached (in the right order) to every gun with a reloadable internal magazine (or we could bloat the regular gun.cs).

The part with the shotgun drum is still true, because we do not want to reload regular shotguns with drums or drum-shotguns with single shells.

**Comment by AutumnNova** [comment]
we could have a box of shotgun shells for some weapons (cshotgun) and just shells for some other weapons (sawnoff), this would create an issue with the shotgun bandolier however

**Comment by drneuris** [comment]
@Tjarr737 
I'm sorry, I hope I don't come off as an asshole, but I see little logic in your post, are you saying "the code is not setup to do things right so i'll do them wrong"? I mean, by all means, but we are talking pretty fundamental mechanics of guns and there will be others working on it too so it'd be wise to start with the right foot.

**Comment by Tjarr737** [comment]
@drneuris My approach was designed to take as little work as possible and change as little outside as possible without compromising on features, with a high chance of remaining compatible even with changes to the magazines itself. I didn't want to just pack the necessary things into Gun.cs because I think adding all the compatibility with this kind of firearm would make it unnecessarily complicated with no real gain.
The code for weapons in general has great limitations and was definitely just designed for guns with magazines, with each type of ammunition having exactly one magazine for it, with any quick fix having bad side effects.
So I take it you are also opposed to idea of a subclass at all?

**Comment by Tjarr737** [comment]
@otherguywolf 
I think making a new category "clips" (stripper clips but also singular bullets/shells) for all types of ammunition designed to reload internal/external magazines would be an adequate solution for that. Then you could reload the internal magazines of non-drum-shotguns with singular shells, shotgun-drums with singular shells and drum-shotguns with drums.
This would require different interaction logic with internal-magazine firearms to process clips instead of magazines.

**Comment by Tjarr737** [comment]
So far everything has been implemented - guns now have a checkbox for containing an internal magazine and there is a new component ClipBehaviour which is used for things like stripper clips or singular cartridges/shells. Shells can be inserted into the shotgun one by one.
I am however currently having a bit of an issue, because the shotgun shells are subvariants of a generic magazine prefab. While I could just add an additional ClipBehaviour on them (as I have done for testing), this is rather unclean.
Does somebody know, how I can solve this subvariant-problem (other than changing Magazines to have a bool for clips and trying to combine both types of behaviour (clips are mostly very simplified magazines without the extra complicated syncing for high speed ammo expenditure)).

Changing the casings should probably be in another overhaul alltogether, where the casings are given by the used ammo type.

Edit: I decided on just making the distinction between magazines and clips with an additional bool instead of actually two different classes. The result is just that clips are more complicated than necessary, but at least there is only one class to manage in the future.

Edit: Magazines and clips are combined now, but there are some network issues to figure out...

Edit: Planned changes from my side done and in PR for now. Still no pump-action or correct casings.
